lib/flowLSP/features/Status/Status.js

```diff
@@ -1,135 +1,140 @@
-      }
+      case 'null':
+        this._widget.setStatus(null);
       default:
-        this.setState({
-          state: null,
-          message: '',
-          progress: ''
-        });
+        this._widget.setStatus(null);
+        break;
     }
   }
 
-  render() {
-    const item = this.statusBarItem;
-    const { state } = this;
-    this._clearTimeout();
+  _getProgress(status: StatusData): string {
+    if (status.kind !== 'yellow') {
+      return '';
+    }
+    if (status.shortMessage != null) {
+      return status.shortMessage;
+    }
+    if (status.progress != null) {
+      const { numerator, denominator } = status.progress;
+      return (
+        Math.round(
+          (numerator / (denominator == null ? 100 : denominator)) * 100,
+        ) + '%'
+      );
+    }
+
+    if (status.message != null) {
+      // remove `Flow:` from message
+      return status.message.replace('Flow:', '');
+    }
 
-    switch (state.state) {
-      case 'running':
-        item.text = 'Flow';
-        item.tooltip = state.message;
-        item.color = this.initialColor;
-        item.show();
+    return '';
+  }
+
+  _handleShowStatus = () => {
+    const { status } = this.state;
+    switch (status.kind) {
+      case 'red': {
+        const items = status.buttons.map(button => ({
+          title: button,
+          command: () => this._provider.clickAction(status.id || '', button),
+        }));
+        vscode.window
+          .showErrorMessage(status.message || '', ...items)
+          .then(selection => {
+            if (selection) {
+              selection.command();
+            }
+          });
         break;
-      case 'busy':
-        item.text = `Flow ${this.spinner()} ${state.progress}`;
-        item.tooltip = state.message;
-        item.color = this.initialColor;
-        item.show();
-        // to run spinner
-        this._timeoutID = setTimeout(() => this.render(), 100);
+      }
+      case 'yellow': {
+        const items = status.buttons.map(button => ({
+          title: button,
+          command: () => this._provider.clickAction(status.id || '', button),
+        }));
+        vscode.window
+          .showWarningMessage(status.message || '', ...items)
+          .then(selection => {
+            if (selection) {
+              selection.command();
+            }
+          });
         break;
-      case 'error':
-        item.text = '$(stop) Flow';
-        item.tooltip = state.message;
-        item.color = 'red';
-        item.show();
+      }
+      case 'green': {
+        vscode.window.showInformationMessage(status.message || '');
         break;
+      }
       default:
-        (state.state: null);
-        item.hide();
-    }
-  }
-
-  _clearTimeout() {
-    if (this._timeoutID) {
-      clearTimeout(this._timeoutID);
-      this._timeoutID = null;
+        break;
     }
-  }
+  };
 
   dispose() {
-    this.statusBarItem.dispose();
-    this._clearTimeout();
+    this._command.dispose();
```

Comment:
can we dispose `StatusProvider` as well, just not to forget about it?

Reply:
@thymikee Actually there is nothing in `StatusProvider` to dispose so I kept it this way.

Reply:
I remember a comment about adding such in StatusProvider. Maybe worth removing it?